bigquery_etl/glean_usage/common.py

```diff
@@ -113,88 +141,113 @@ def _extract_dataset_from_glob(pattern):
+            project_id=project_id,
         )
-        if args.output_only:
+
+        render_kwargs.update(self.custom_render_kwargs)
+        render_kwargs.update(tables)
+
+        query_sql = render(query_filename, **render_kwargs)
+        view_sql = render(view_filename, **render_kwargs)
+        view_metadata = render(view_metadata_filename, format=False, **render_kwargs)
+
+        if not self.no_init:
+            try:
+                init_sql = render(init_filename, **render_kwargs)
+            except TemplateNotFound:
+                init_sql = render(query_filename, init=True, **render_kwargs)
+
+        if not (referenced_table_exists(view_sql)):
+            logging.info(
+                "Skipping view for table which doesn't exist:"
+                f" {self.target_table_id}"
+            )
             return
-        logging.info(f"Dry runs successful for {len(baseline_tables)} table(s)")
-        # Now, actually run the queries.
-        if not args.dry_run:
-            pool.map(
-                partial(
-                    run_query_callback, args.project_id, date=args.date, dry_run=False
-                ),
-                baseline_tables,
+
+        if output_dir:
+            write_sql(output_dir, view, "metadata.yaml", view_metadata)
+            write_sql(output_dir, view, "view.sql", view_sql)
+            write_sql(output_dir, table, "query.sql", query_sql)
+
+            if not self.no_init:
+                write_sql(output_dir, table, "init.sql", init_sql)
+
+            write_dataset_metadata(output_dir, view)
+
+    def generate_per_app(self, project_id, app_info, output_dir=None):
+        """Generate the baseline table query per app_name."""
+        target_view_name = "_".join(self.target_table_id.split("_")[:-1])
+        target_dataset = app_info[0]["app_name"]
+
+        datasets = [
+            (a["bq_dataset_family"], a["app_channel"])
+            for a in app_info
+            if "app_channel" in a
+        ]
+
+        # Some apps only have a single channel, in which case the per-app_id dataset
+        # is already sufficient.
+        if len(datasets) == 0:
+            return
+
+        render_kwargs = dict(
+            header="-- Generated via bigquery_etl.glean_usage\n",
+            project_id=project_id,
+            target_view=f"{target_dataset}.{target_view_name}",
+            datasets=datasets,
+            table=target_view_name,
+        )
+
+        sql = render("cross_channel.view.sql", **render_kwargs)
+
+        if not (referenced_table_exists(sql)):
+            logging.info(
+                "Skipping view for table which doesn't exist:"
+                f" {self.target_table_id}"
             )
+            return
+
+        view = f"{project_id}.{target_dataset}.{target_view_name}"
+
+        if output_dir:
+            write_sql(output_dir, view, "view.sql", sql)
+            write_dataset_metadata(output_dir, view)
```

Comment:
We need to write out the dataset metadata before we do the referenced_table_exists check. `referenced_table_exists` fails due to the per-app dataset not yet existing.

Reply:
We also may need to create the directory in write_dataset_metadata.